### PR TITLE
Xla const output cache on GPU 

### PR DESCRIPTION
**Pull request by zhaozheng09**
#53609

**Problem overview:** the const outputs of xla executable is kept in host memory. if the const output of executable need a device tensor, we need to do an H2D(host memory to device memory) .

**Ideas overview:** cache the const output of xla executable to device memory, then replace H2D with lookup from cache.

**Background(features async H2D)** :

The total ops in out model are set on the gpu device, so model spends data very fast, but the data part(parsing the data from TFRecord to Features) complete run on host memory, so we need copy features from host to device, this H2D will launch lots of requests of H2D on H2D stream. we use prefetch_to_device to overlap the time required for feature H2D.(prefetch_to_device: async copy features from host to device. The method of model obtaining data is replaced from synchronous H2D to obtain from cache)

**Problem**:

We use xla to increase speed of training, but we found gpu util is 0 after xla run and model hung, As follows:

We carefully analyse nsys profiling file, found xla brought H2D once, and this H2D will block by features H2D in background.

**Solution**:

We found that this H2D is generated by the xla op, and then we read the code about xla carefully, we found that the const output tensor of the executable is placed on the host, and then copy to the device side through memcpyH2D.

We can add a cache<executable output index, device tensor> on device. We put pair<executable output index, device tensor> to cache in the first time, and the next time replace to get device tensor from synchronous H2D to get device tensor from cache.

**result(special model)**:

no prefetch_to_device: No performance drop.

prefetch_to_device: preformance 15%↑

Is there a pr template？ It's a little late in China, can I change it tomorrow?

### CODE REVIEW

**Comment by zhaozheng09** [comment]
https://github.com/tensorflow/tensorflow/issues/53609

**Comment by joker-eph** [comment]
(tagged so that we run some internal tests, still pending more code review)

**Comment by cheshire** [comment]
@paynecl

**Comment by zhaozheng09** [comment]
Why does this check fail ? and Details is ```http://cl/423940400``` ???

**Comment by cheshire** [comment]
I've tagged it manually I think. Could you write a larger PR description?

File: tensorflow/compiler/jit/kernels/xla_ops.cc [hunk 1]
```diff
@@ -307,11 +321,26 @@ void XlaLocalLaunchBase::Compute(OpKernelContext* ctx) {
 
   auto elapsed = env->NowMicros() - start_time;
   VLOG(2) << "Elapsed time: " << elapsed << "us";
+
+  if (cache_ == nullptr) {
+    rm_ = ctx->resource_manager();
+    rm_->LookupOrCreate<XlaConstantOutputResource>(
```
Comment (akuegel) [inline tensorflow/compiler/jit/kernels/xla_ops.cc:327]: Compiler error: ignoring return value of function declared with 'warn_unused_result' attribute [-Werror,-Wunused-result]

I think you may need a TF_RETURN_IF_ERROR macro around this call.

**Review by akuegel — APPROVED** [review]

**Comment by zhaozheng09** [comment]
ci test:
AMD ROCm: ```No module named 'packaging'```
window error: ```404 Not Found```

**Comment by akuegel** [comment]
Your PR was rolled back because it caused data race issues (detected with TSAN). I believe you may need to add locks when accessing the cache.

**Comment by zhaozheng09** [comment]
> Your PR was rolled back because it caused data race issues (detected with TSAN). I believe you may need to add locks when accessing the cache.



> Your PR was rolled back because it caused data race issues (detected with TSAN). I believe you may need to add locks when accessing the cache.

I‘m sorry. 
How can I reproduce this problem or screenshot of TSAN report.

**Comment by zhaozheng09** [comment]
> Your PR was rolled back because it caused data race issues (detected with TSAN). I believe you may need to add locks when accessing the cache.

I'm so sorry. 
We use this feature on many of our models and have seen no data races, so can you give me a reproduction environment or a TSAN report or core dump file to help me to solve this data races?

File: tensorflow/compiler/jit/kernels/xla_ops.cc [hunk 2]
```diff
@@ -562,12 +605,26 @@ void XlaRunOp::Compute(OpKernelContext* ctx) {
       ctx, *closure.compilation_result(), closure.num_constant_args());
   OP_REQUIRES_OK(ctx, variable_infos.status());
   OP_REQUIRES_OK(ctx, LockVariables(absl::MakeSpan(*variable_infos)));
+
+  if (cache_ == nullptr) {
```
Comment (akuegel) [inline tensorflow/compiler/jit/kernels/xla_ops.cc:609]: The race seems to be between this line and the assignment of cache_ in rm_->LookupOrCreate.
So I think you need a mutex lock here around this block (and I guess everywhere where cache_ is accessed).

**Comment by akuegel** [comment]
> I'm so sorry. We use this feature on many of our models and have seen no data races, so can you give me a reproduction environment or a TSAN report or core dump file to help me to solve this data races?
> 
> Maybe two session run a graph ?

The test which fails is a TPU test, so unfortunately I cannot give you a reproduction environment. But I added a comment to the code which code lines are involved in the TSAN report finding.

**Comment by zhaozheng09** [comment]
> > I'm so sorry. We use this feature on many of our models and have seen no data races, so can you give me a reproduction environment or a TSAN report or core dump file to help me to solve this data races?
> > Maybe two session run a graph ?
> 
> The test which fails is a TPU test, so unfortunately I cannot give you a reproduction environment. But I added a comment to the code which code lines are involved in the TSAN report finding.

I probably know where the problem is, and I will test on our model .

**Pull request merged**